druid/src/core.rs

```diff
@@ -335,25 +344,32 @@ impl<T: Data, W: Widget<T>> WidgetPod<T, W> {
     /// [`paint`]: trait.Widget.html#tymethod.paint
     /// [`paint_with_offset`]: #method.paint_with_offset
     pub fn paint(&mut self, ctx: &mut PaintCtx, data: &T, env: &Env) {
+        // we need to do this before we borrow from self
+        if env.get(Env::DEBUG_WIDGET_ID) {
+            self.make_widget_id_layout_if_needed(self.state.id, ctx, env);
+        }
+
         let mut inner_ctx = PaintCtx {
             render_ctx: ctx.render_ctx,
             window_id: ctx.window_id,
             z_ops: Vec::new(),
             region: ctx.region.clone(),
             base_state: &self.state,
             focus_widget: ctx.focus_widget,
+            depth: ctx.depth,
         };
         self.inner.paint(&mut inner_ctx, data, env);
-        ctx.z_ops.append(&mut inner_ctx.z_ops);
 
-        if env.get(Env::DEBUG_PAINT) {
-            const BORDER_WIDTH: f64 = 1.0;
-            let rect = inner_ctx.size().to_rect().inset(BORDER_WIDTH / -2.0);
-            let id = self.id().to_raw();
-            let color = env.get_debug_color(id);
-            inner_ctx.stroke(rect, &color, BORDER_WIDTH);
-        }
+        let debug_layout = env.get(Env::DEBUG_PAINT);
+        let debug_ids = env.get(Env::DEBUG_WIDGET_ID) && inner_ctx.is_hot();
 
+        if debug_layout {
+            self.debug_paint_layout_bounds(&mut inner_ctx, env);
+        }
+        if debug_ids {
+            self.debug_paint_widget_ids(&mut inner_ctx, env);
+        }
```

Comment:
The `debug_paint_widget_ids` method might call `debug_paint_layout_bounds`, so I think we can optimize it into something like this:
```rust
// Check the bool first, because most widgets aren't hot and we won't need to do the map lookup
let debug_ids = inner_ctx.is_hot() && env.get(Env::DEBUG_WIDGET_ID);

let debug_layout = if debug_ids {
    // Change debug_paint_widget_ids to return true if it called debug_paint_layout_bounds.
    // Then we won't need to call it again or even do the env map lookup.
    !self.debug_paint_widget_ids(&mut inner_ctx, env) && env.get(Env::DEBUG_PAINT)
} else {
    env.get(Env::DEBUG_PAINT)
};
if debug_layout {
    self.debug_paint_layout_bounds(&mut inner_ctx, env);
}
```

Reply:
okay, I think we can assume that debug_paint_widget_ids always paints the layout so I can simplify this a bit further, overall looks reasonable though.